### ISSUE TITLE
Deadlock during NewSession

### ISSUE THREAD

**Issue by kthommandra** [gocql/gocql#928]
Found on da4a50306e7bb46de0d077d7ac78c297de9aaac4

As part of NewSession, ring information is collected using some queries on system tables. If there is a query error then it leads to deadlock as shown in the stack trace below

[stack.txt](https://github.com/gocql/gocql/files/1110740/stack.txt)

**Comment by kthommandra** [comment]
I hit this issue in our setup and was able to recreate the issue by reducing the connection timeout value after setting up the control connection and  

```
        if !s.cfg.DisableInitialHostLookup {
            {
                // Change the cnxn timeout to an extremely low
                // value after the control cnxn has been setup
                conn := s.control.conn.Load().(*Conn)
                conn.timeout = 1 * time.Nanosecond
            }

            var partitioner string
```

**Comment by kthommandra** [comment]
Here's my analysis of the issue ... appreciate comments.

1) errors during establishing the control conn (controlConn.withConn), cause reconnect without ring refresh
2) errors during heartbeat (controlConn.heartBeat) causes reconnect with ring refresh
3) errors in query execution on control connection (controlConn.HandleError) cause reconnect with ring refresh

- during query error on control conn, the controlConn.HandleError is invoked before the reconnect in controlConn.withConn

- GetHosts (part of ring refresh) is not meant to be re-entrant (acquires ringDescriber mutex at the start) and in general it doesn't make sense for it to be re-entrant.

- GetHosts uses queries (controlConn.withConn) and errors during those queries can cause ring refresh to be invoked causing re-entry into GetHost and hence deadlock

- To avoid ring refresh in controlConn.query, 1 is not sufficient, 3 should be avoided too.

- Also, in reconnect(true) code path, the Conn is setup with default error handler (3), and during refresh ring, the query failure can cause reconnect to be re-entered.

- I couldn't find a specific scenario where the controlConn.HandleError doing the ring refresh helps. I see it more as a means of implementing intent (3)

One way to fix this seems to be:

- disable the ring refresh in handleError, so do reconnect(false)
- have two APIs controlConn.query and controlConn.queryRefreshRing - the queries which are issued from GetHosts code path will use .query while rest of the queries on control cnxn can use .queryRefreshRing. This way we maintain the intent of - refresh ring after re-connection in general case while avoiding the ring refresh for special cases.

What are your thoughts? I can send a PR once the strategy emerges.

**Comment by Zariel** [comment]
Thanks for the investigation, I think what should happen is if 2 goroutines attempt to reconnect the control connection that the first one wins and the second waits for the result of the reconnection.

**Comment by kthommandra** [comment]
I'm not sure if this is a multi-go routine issue.

Given that, reconnect(true) involves query (for ring refresh), the controlConn.HandleError doing a reconnect(true) sounds incorrect given ring refresh is non-reentrant.
Change #504 made HandleError use reconnect(true). 

Do you see an issue if we do the following:

- heartBeat is the only codepath where reconnect(true) is done and hence ring is refreshed in one place only. The heartBeat will be changed to notice that the controlConn has changed and do a ring refresh apart from error path. I think ring information getting updated in background is fine.
- controlConn.HandleError does a reconnect(false) thus a query error will at least cause the next query attempt to use a new cnxn (as in controlConn.query code path)
- withConn does reconnect(false) , no change here

I'll try out these changes.

**Comment by kthommandra** [comment]
Just to clarify - IMHO, the deadlock issue discussed here isn't related to multiple go-routines. I agree that reconnect needs to be multi-goroutine safe. However, thats a separate issue.

**Comment by dankinder** [comment]
I appear to be hitting some kind of deadlock as well. Having just recently upgraded from go 1.7 to 1.8 may have made a difference. Here are the gocql-relevant goroutine stacks.

```
677 @ 0x42e19a 0x4292b7 0x4288f9 0x5374e8 0x537554 0x538ca7 0x54c2f0 0x50bd92 0x463339 0x4634a8 0x850f48 0x842c2a 0x84289b 0x45b531
#	0x4288f8	net.runtime_pollWait+0x58								/usr/local/go/src/runtime/netpoll.go:164
#	0x5374e7	net.(*pollDesc).wait+0x37								/usr/local/go/src/net/fd_poll_runtime.go:75
#	0x537553	net.(*pollDesc).waitRead+0x33								/usr/local/go/src/net/fd_poll_runtime.go:80
#	0x538ca6	net.(*netFD).Read+0x1b6									/usr/local/go/src/net/fd_unix.go:250
#	0x54c2ef	net.(*conn).Read+0x6f									/usr/local/go/src/net/net.go:181
#	0x50bd91	bufio.(*Reader).Read+0x311								/usr/local/go/src/bufio/bufio.go:213
#	0x463338	io.ReadAtLeast+0xa8									/usr/local/go/src/io/io.go:307
#	0x4634a7	io.ReadFull+0x57									/usr/local/go/src/io/io.go:325
#	0x850f47	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.readHeader+0x97	/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/frame.go:364
#	0x842c29	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*Conn).recv+0xb9	/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/conn.go:466
#	0x84289a	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*Conn).serve+0x2a	/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/conn.go:428
```
```
4 @ 0x42e19a 0x43d594 0x43c1fc 0x84e646 0x45b531
#	0x84e645	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*eventDebouncer).flusher+0x135	/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/events.go:39

2 @ 0x42e19a 0x42e27e 0x43ea81 0x43e684 0x462969 0x864dfc 0x84ff2a 0x84f6e7 0x89d0a8 0x45b531
#	0x43e683	sync.runtime_Semacquire+0x33																			/usr/local/go/src/runtime/sema.go:47
#	0x462968	sync.(*RWMutex).RLock+0x48																			/usr/local/go/src/sync/rwmutex.go:43
#	0x864dfb	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*HostInfo).Version+0x4b											/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/host_source.go:292
#	0x84ff29	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*Session).handleNodeUp+0x99											/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/events.go:245
#	0x84f6e6	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*Session).handleNodeEvent+0x716										/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/events.go:169
#	0x89d0a7	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*Session).(ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.handleNodeEvent)-fm+0x47	/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/session.go:115

2 @ 0x42e19a 0x42e27e 0x43ea81 0x43e764 0x461a9d 0x88c69d 0x88d127 0x84fbca 0x84f852 0x89d0a8 0x45b531
#	0x43e763	sync.runtime_SemacquireMutex+0x33																		/usr/local/go/src/runtime/sema.go:62
#	0x461a9c	sync.(*Mutex).Lock+0x9c																				/usr/local/go/src/sync/mutex.go:87
#	0x88c69c	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*cowHostList).add+0x3c											/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/policies.go:46
#	0x88d126	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*tokenAwareHostPolicy).AddHost+0x36										/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/policies.go:306
#	0x84fbc9	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*Session).handleNewNode+0x309										/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/events.go:200
#	0x84f851	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*Session).handleNodeEvent+0x881										/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/events.go:162
#	0x89d0a7	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*Session).(ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.handleNodeEvent)-fm+0x47	/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/session.go:115
```
```
1 @ 0x42e19a 0x42e27e 0x43ea81 0x43e684 0x462969 0x864258 0x863c4e 0x88c855 0x88d127 0x868e20 0x84fc1c 0x8500d3 0x84cda1 0x84caed 0x890f89 0x89069f 0x8401e2 0x8bd28a 0xa21919 0xa16119 0xa16b09 0xa167ab 0xa283b5 0x42dd4a 0x45b531
#	0x43e683	sync.runtime_Semacquire+0x33										/usr/local/go/src/runtime/sema.go:47
#	0x462968	sync.(*RWMutex).RLock+0x48										/usr/local/go/src/sync/rwmutex.go:43
#	0x864257	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*HostInfo).ConnectAddress+0x57	/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/host_source.go:182
#	0x863c4d	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*HostInfo).Equal+0x7d		/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/host_source.go:135
#	0x88c854	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*cowHostList).add+0x1f4		/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/policies.go:54
#	0x88d126	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*tokenAwareHostPolicy).AddHost+0x36	/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/policies.go:306
#	0x868e1f	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*ringDescriber).refreshRing+0x1df	/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/host_source.go:686
#	0x84fc1b	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*Session).handleNewNode+0x35b	/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/events.go:203
#	0x8500d2	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*Session).handleNodeUp+0x242	/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/events.go:255
#	0x84cda0	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*controlConn).setupConn+0x1a0	/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/control.go:251
#	0x84caec	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*controlConn).connect+0x18c		/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/control.go:229
#	0x890f88	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*Session).init+0x358		/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/session.go:179
#	0x89069e	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.NewSession+0x86e			/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/session.go:143
#	0x8401e1	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*ClusterConfig).CreateSession+0x61	/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/cluster.go:153
#	0x8bd289	ghe.iparadigms.com/iParadigms/walker/cassandra.NewDispatcher+0x99					/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/cassandra/dispatcher.go:108
#	0xa21918	ghe.iparadigms.com/iParadigms/walker/cmd.NewCommander.func6+0x328					/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/cmd/cmd.go:293
#	0xa16118	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/spf13/cobra.(*Command).execute+0x298		/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/spf13/cobra/command.go:653
#	0xa16b08	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/spf13/cobra.(*Command).ExecuteC+0x338		/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/spf13/cobra/command.go:728
#	0xa167aa	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/spf13/cobra.(*Command).Execute+0x2a		/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/spf13/cobra/command.go:687
#	0xa283b4	main.main+0x1d4												/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/walker/main.go:69
#	0x42dd49	runtime.main+0x209			

1 @ 0x42e19a 0x42e27e 0x43ea81 0x43e684 0x462a8e 0x864f5a 0x85017f 0x45b531
#	0x43e683	sync.runtime_Semacquire+0x33										/usr/local/go/src/runtime/sema.go:47
#	0x462a8d	sync.(*RWMutex).Lock+0x6d										/usr/local/go/src/sync/rwmutex.go:91
#	0x864f59	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*HostInfo).setState+0x39		/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/host_source.go:311
#	0x85017e	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*Session).handleNodeDown+0x8e	/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/events.go:272

1 @ 0x42e19a 0x42e27e 0x43ea81 0x43e684 0x462a8e 0x864f5a 0x85017f 0x84f6ab 0x89d0a8 0x45b531
#	0x43e683	sync.runtime_Semacquire+0x33																			/usr/local/go/src/runtime/sema.go:47
#	0x462a8d	sync.(*RWMutex).Lock+0x6d																			/usr/local/go/src/sync/rwmutex.go:91
#	0x864f59	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*HostInfo).setState+0x39											/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/host_source.go:311
#	0x85017e	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*Session).handleNodeDown+0x8e										/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/events.go:272
#	0x84f6aa	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*Session).handleNodeEvent+0x6da										/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/events.go:171
#	0x89d0a7	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*Session).(ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.handleNodeEvent)-fm+0x47	/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/session.go:115

1 @ 0x42e19a 0x42e27e 0x43ea81 0x43e764 0x461a9d 0x462a4d 0x865161 0x84fc31 0x84f852 0x89d0a8 0x45b531
#	0x43e763	sync.runtime_SemacquireMutex+0x33																		/usr/local/go/src/runtime/sema.go:62
#	0x461a9c	sync.(*Mutex).Lock+0x9c																				/usr/local/go/src/sync/mutex.go:87
#	0x462a4c	sync.(*RWMutex).Lock+0x2c																			/usr/local/go/src/sync/rwmutex.go:86
#	0x865160	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*HostInfo).update+0x30											/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/host_source.go:344
#	0x84fc30	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*Session).handleNewNode+0x370										/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/events.go:195
#	0x84f851	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*Session).handleNodeEvent+0x881										/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/events.go:162
#	0x89d0a7	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*Session).(ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.handleNodeEvent)-fm+0x47	/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/session.go:115

1 @ 0x42e19a 0x42e27e 0x43ea81 0x43e764 0x461a9d 0x88ca1b 0x88d1cd 0x88d295 0x8501a0 0x45b531
#	0x43e763	sync.runtime_SemacquireMutex+0x33										/usr/local/go/src/runtime/sema.go:62
#	0x461a9c	sync.(*Mutex).Lock+0x9c												/usr/local/go/src/sync/mutex.go:87
#	0x88ca1a	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*cowHostList).remove+0x4a			/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/policies.go:97
#	0x88d1cc	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*tokenAwareHostPolicy).RemoveHost+0x5c	/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/policies.go:313
#	0x88d294	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*tokenAwareHostPolicy).HostDown+0x34	/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/policies.go:324
#	0x85019f	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*Session).handleNodeDown+0xaf		/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/events.go:273

1 @ 0x42e19a 0x43d594 0x43c1fc 0x84bceb 0x45b531
#	0x84bcea	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*controlConn).heartBeat+0x2ba	/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/control.go:63

1 @ 0x42e19a 0x43d594 0x43c1fc 0x891481 0x45b531
#	0x891480	ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql.(*Session).reconnectDownedHosts+0x240	/var/lib/jenkins/jobs/Walker/jobs/walker-master-build/gopath/src/ghe.iparadigms.com/iParadigms/walker/vendor/github.com/gocql/gocql/session.go:228
```

**Comment by kthommandra** [comment]
Pull request #998 does not fix issue #928

See PR #1015 

Issue #928 is caused by a deadlock when the query used in GetHosts code path hits an error.
Two scenarios:
a) GetHosts -> getClusterPeerInfo -> conn.exec -> closeWithError -> control.HandleError -> reconnect -> ring.refresh -> GetHosts
b) GetHosts -> getClusterPeerInfo -> conn.exec -> handleTimeout -> closeWithError -> control.HandleError -> reconnect -> ring.refresh -> GetHosts

[scenario-a.txt](https://github.com/gocql/gocql/files/1436588/scenario-a.txt)
[scenario-b.txt](https://github.com/gocql/gocql/files/1436589/scenario-b.txt)

We hit scenario (b) in our production system when some of the nodes were taken down. The issue happened on a couple of systems running GoCQL based app. Refer to scenario-b.txt attachment.

Scenario (a) can be recreated using patch scenario-a.txt
Note: the patch does few things:

disabled use of SSL (unrelated to this change, merely a workaround for some SSL issue in my setup)
reduce the timeout on the control connection after the connection is established to force the query as part of GetHosts to fail in conn.exec with closeWithError (scenario-a)
change the TestCreateSessionTimeout to ensure the session setup is attempted
added some instrumentation to catch the deadlock in GetHosts
disable unit tests and other misc changes in integration.sh to enable a quick recreate
There are a couple of ways of fixing this deadlock but restricting the ring refresh to heartBeat seems like the cleanest and simplest fix. I couldn't find anything that depends on the ring to be refreshed as part of the control connection's error handler.
